src/PdoConnection.php

```diff
@@ -238,7 +252,13 @@ public function begin()
      */
     public function rollBack()
     {
-        return $this->connect()->rollBack();
+        if (--$this->transactionCount === 0) {
+            $this->rollbackNeeded = false;
+            return $this->connect()->rollBack();
+        }
+
+        $this->rollbackNeeded = true;
```

Comment:
The flag is false by default and is only set to true when a rollback is called on an inner transaction.  It doesn't depend on the current value of rollbackNeeded at all.